### PR TITLE
gRPC

### PR DESCRIPTION
**Pull request by jendakol**
This is still WIP because it has not been tested with the Authenticator component yet. However, I believe it will be possible and if not, it will require only minor changes so please do the review.  
I will try to provide a readme - you can check `ClientTest` and `ServerTest` for usage until then.

### CODE REVIEW

**Comment by jendakol** [comment]
Example of generated code:
```scala
package com.avast.cactus.grpc.server

import com.avast.cactus.grpc._
import com.avast.cactus.grpc.server.TestApi.{GetRequest, GetResponse}
import com.avast.cactus.grpc.server.TestApiServiceGrpc.{TestApiServiceFutureStub, TestApiServiceImplBase}
import com.avast.cactus.v3._
import io.grpc.ForwardingClientCall.SimpleForwardingClientCall
import io.grpc._
import io.grpc.inprocess.{InProcessChannelBuilder, InProcessServerBuilder}
import io.grpc.stub.StreamObserver
import org.scalactic.{Bad, Good}

import scala.collection.JavaConverters._
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.Try

case class MyRequest(names: Seq[String])

case class MyContext(theHeader: String, headerFromOwnInterceptor: Int)

case class MyResponse(results: Map[String, Int])

trait MyApi {
  def get(request: MyRequest, ctx: MyContext): Future[Either[Status, MyResponse]]
}

object Test extends App {

  val metadataKey = Metadata.Key.of(UserHeaderPrefix + "theHeader", Metadata.ASCII_STRING_MARSHALLER)

  val impl = new MyApi {
    override def get(request: MyRequest, ctx: MyContext): Future[Either[Status, MyResponse]] = {
      Future.successful(Right(MyResponse(Map("str" -> 42))))
    }
  }

  import ServerCommonMethods._

  /* GENERATED BY MACRO START */
  
  val mappedService = new TestApiServiceImplBase {
    private val interceptorsWrapper = new ServerInterceptorsWrapper(
      List(
        new ServerAsyncInterceptor {
          override def apply(m: GrpcMetadata): Future[Either[Status, GrpcMetadata]] = Future.successful {
            Right {
              m.copy(context = {
                val ctx = m.context
                ctx.withValue(ContextKeys.get[Int]("headerFromOwnInterceptor"), 42)
              })
            }
          }
        }
      ))

    override def get(request: GetRequest, responseObserver: StreamObserver[GetResponse]): Unit = {
      def createCtx: Try[MyContext] = {
        {
          val ctx = Context.current()

          for {
            theHeader <- Option(ContextKeys.get[String]("theHeader").get(ctx))
            headerFromOwnInterceptor <- Option(ContextKeys.get[Int]("headerFromOwnInterceptor").get(ctx))
          } yield {
            MyContext(theHeader, headerFromOwnInterceptor)
          }
        }.map(scala.util.Success(_))
          .getOrElse {
            scala.util.Failure(new IllegalArgumentException("Missing headers for creation of the context"))
          }
      }

      val result = interceptorsWrapper.withInterceptors {
        withContext[MyRequest, GetResponse, MyContext](createCtx) { ctx =>
          request.asCaseClass[MyRequest] match {
            case Good(req) => executeRequest[MyRequest, MyResponse, GetResponse](req, impl.get(_, ctx))
            case Bad(errors) =>
              Future.successful {
                Left {
                  new StatusException(Status.INVALID_ARGUMENT.withDescription(formatCactusFailures("request", errors)))
                }
              }
          }
        }
      }

      result
        .recover(recoverWithStatus)
        .andThen(sendResponse(responseObserver))
    }
  }

  /* GENERATED BY MACRO END */

  InProcessServerBuilder
    .forName("channelName")
    .directExecutor
    .addService(
      ServerInterceptors.intercept(
        mappedService,
        ServerMetadataInterceptor
      ))
    .build
    .start

  val stub: TestApiServiceFutureStub = {
    val channel = InProcessChannelBuilder.forName("channelName").directExecutor.build
    TestApiServiceGrpc.newFutureStub(channel).withInterceptors(ClientMetadataInterceptorTest)
  }

  val result = stub.get(GetRequest.newBuilder().build()).get()

  println(result)

  private[grpc] object ClientMetadataInterceptorTest extends ClientInterceptor {
    override def interceptCall[ReqT, RespT](method: MethodDescriptor[ReqT, RespT],
                                            callOptions: CallOptions,
                                            next: Channel): ClientCall[ReqT, RespT] = {

      new SimpleForwardingClientCall[ReqT, RespT](next.newCall(method, callOptions)) {
        override def delegate(): ClientCall[ReqT, RespT] = super.delegate()

        override def start(responseListener: ClientCall.Listener[RespT], headers: Metadata): Unit = {
          headers.put(metadataKey, s"theHeader-theValue")
          super.start(responseListener, headers)
        }
      }
    }
  }

}

```

**Comment by jendakol** [comment]
I have added the README and also testing class which demonstrates the code generated by macros. @augi @petr-k

File: grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientInterceptorsWrapper.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                  clientCall
+                }
+              })
+
+          case Left(statusException) => Future.failed(statusException)
+        }
+        .recover {
+          case e: StatusException => Left(ServerError(e.getStatus))
+          case e: StatusRuntimeException => Left(ServerError(e.getStatus))
+          case NonFatal(e) => Left(ServerError(Status.ABORTED.withCause(e).withDescription("Request could not been processed")))
```
Comment (petr-k) [inline grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientInterceptorsWrapper.scala]: _"...could not be processed"_

File: grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientMacros.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        }
+       """
+    }
+  }
+
+  private def generateMappingMethod(implMethod: ImplMethod, apiMethod: ApiMethod): Tree = {
+    q"""
+       override def ${implMethod.name}(request: ${implMethod.request}): scala.concurrent.Future[com.avast.cactus.grpc.ServerResponse[${implMethod.response}]] = withInterceptors {
+          request.asGpb[${apiMethod.request}] match {
+             case org.scalactic.Good(req) => executeRequest[${apiMethod.request}, ${apiMethod.response}, ${implMethod.response}](req, stub.${apiMethod.name}, ex)
```
Comment (petr-k) [inline grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientMacros.scala]: Just out of curiosity, is there a reason to use both `scalactic` and `cats` in one library?

File: grpc-common/README.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      override def apply(m: GrpcMetadata): Future[Either[Status, GrpcMetadata]] = {
+        val cont = MyContextContent(42, "jenda")
+        
+        Future.successful {
+          Right(m.copy(context = m.context.withValue(ContextKeys.get[MyContextContent]("content"), cont)))
+        }
+      }
+    }
+    ```
+Note that names of `MyContext` class fields have to match keys in `Context`. If some field is missing, an `INVALID_ARGUMENT` status is
```
Comment (petr-k) [inline grpc-common/README.md]: If the context case class has an `Option` parameter and the key is missing, will the `INVALID_ARGUMENT` status be still returned?

File: grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientMacros.scala (hunk 2 above)
Comment (jendakol) [reply grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientMacros.scala]: I am thinking about changing the API to `Either` from `scalactic.Or`, as well as compiling only for Scala 2.12. That change could mean also removing the Scalactic completely if I will implement the errors accumulation with Cats.
However that's a subject to another PR.

File: grpc-common/README.md (hunk 3 above)
Comment (jendakol) [reply grpc-common/README.md]: Good question. Currently the context class mapping does NOT support `Option`s (when mapping headers).  Do you think it should support it?

File: grpc-common/src/main/scala/com/avast/cactus/grpc/ContextKeys.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package com.avast.cactus.grpc
+
+import io.grpc.Context
+import io.grpc.Context.Key
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object ContextKeys {
```
Comment (augi) [inline grpc-common/src/main/scala/com/avast/cactus/grpc/ContextKeys.scala]: Please copy shorter version of [this paragraph](https://github.com/avast/cactus/blob/gRPC/grpc-common/README.md#providing-own-interceptors) to the comment.

File: grpc-common/README.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+val ctx = Context.current()
+
+for {
+  theHeader <- Option(ContextKeys.get[String]("theHeader").get(ctx))
+  content <- Option(ContextKeys.get[MyContextContent]("content").get(ctx))
+} yield {
+  MyContext(theHeader, content)
+}
+```
+1. `theHeader` is header received from the client. Headers are automatically passed to the `Context` to be available for this - no action required
```
Comment (augi) [inline grpc-common/README.md]: What data types are supported for client headers? String only? If so, please mention it.

Comment (jendakol) [reply grpc-common/README.md]: I will.

File: grpc-common/src/main/scala/com/avast/cactus/grpc/ContextKeys.scala (hunk 4 above)
Comment (jendakol) [reply grpc-common/src/main/scala/com/avast/cactus/grpc/ContextKeys.scala]: OK

File: grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientInterceptorsWrapper.scala (hunk 1 above)
Comment (jendakol) [reply grpc-client/src/main/scala/com/avast/cactus/grpc/client/ClientInterceptorsWrapper.scala]: ;-) Refactoring..

File: grpc-common/README.md (hunk 3 above)
Comment (petr-k) [reply grpc-common/README.md]: Not necessarily, but would be nice to have. Could be a future improvement.

**Pull request merged**